File: library/src/main/java/org/incendo/interfaces/grid/GridMap.java
package org.incendo.interfaces.grid;

import java.util.Map;

public interface GridMap<V> extends Map<GridPoint, V> {

    default void put(int row, int column, V value) {
        this.put(GridPoint.at(row, column), value);
    }

    default V get(int row, int column) {
        return this.get(GridPoint.at(row, column));
    }

    default boolean containsKey(int row, int column) {
        return this.containsKey(GridPoint.at(row, column));
    }
}

File: next/src/main/java/org/incendo/interfaces/next/grid/GridMap.kt
package org.incendo.interfaces.next.grid

public interface GridMap<V> {

    public operator fun set(row: Int, column: Int, value: V)

    public operator fun set(vector: GridPoint, value: V) {
        set(vector.x, vector.y, value)
    }

    public operator fun get(row: Int, column: Int): V?

    public operator fun get(vector: GridPoint): V? {
        return get(vector.x, vector.y)
    }

    public fun has(row: Int, column: Int): Boolean

    public fun has(vector: GridPoint): Boolean {
        return has(vector.x, vector.y)
    }

    public fun forEach(consumer: (row: Int, column: Int, V) -> Unit)

    public suspend fun forEachSuspending(consumer: suspend (row: Int, column: Int, V) -> Unit)
}
